## 1. Import Required Libraries

In [2]:
import pandas as pd
import numpy as np

## 2. Load the Dataset

In [3]:
df = pd.read_csv("../00-dataSource/raw_data.csv")

## 3. Initial Data Exploration

In [4]:
# Show first 5 rows: quick preview of dataset structure and values
print("First 5 rows:")
display(df.head())

# Show dataset size: number of rows and columns
print("Shape of dataset:")
print(df.shape)

# Show column names: list of all features
print("\nColumn names:")
print(df.columns)

# Show index: how rows are labeled (usually default range)
print("\nIndex:")
print(df.index)

# Show numerical summary: mean, min, max, etc.
print("\nStatistical summary (Numerical):")
display(df.describe())

# Show data types and non-null counts: detect missing values & wrong types
print("\nDataset information:")
df.info()

# Show missing values per column: identify incomplete data
print("\nMissing values:")
display(df.isnull().sum())

# Show number of duplicate rows: detect repeated records
print("\nDuplicate rows:")
print(df.duplicated().sum())

First 5 rows:


,id,age,gender,region,income_level,education_level,daily_role,device_hours_per_day,phone_unlocks,notifications_per_day,...,sleep_quality,anxiety_score,depression_score,stress_level,happiness_score,focus_score,high_risk_flag,device_type,productivity_score,digital_dependence_score
0,1,40,Female,Asia,High,High School,Part-time/Shift,3.54,45,561,...,3.353627,9.926651,5.0,6.593289,8.0,23.0,0,Android,70.000000,25.700000
1,2,27,Male,Africa,Lower-Mid,Master,Full-time Employee,5.65,100,393,...,2.908147,4.000000,4.0,4.126926,8.1,35.0,0,Laptop,64.000000,30.100000
2,3,31,Male,North America,Lower-Mid,Bachelor,Full-time Employee,8.87,181,231,...,2.889213,4.000000,8.0,1.429139,7.6,15.0,0,Android,65.299301,40.600000
3,4,41,Female,Middle East,Low,Master,Caregiver/Home,4.05,94,268,...,3.097488,7.093357,9.0,4.995512,7.8,28.0,1,Tablet,80.000000,36.684152
4,5,26,Female,Europe,Lower-Mid,Bachelor,Full-time Employee,13.07,199,91,...,2.786098,7.028125,15.0,9.448757,4.2,70.0,1,Android,65.299301,48.400000


Shape of dataset:
(3500, 24)

Column names:
Index(['id', 'age', 'gender', 'region', 'income_level', 'education_level',
       'daily_role', 'device_hours_per_day', 'phone_unlocks',
       'notifications_per_day', 'social_media_mins', 'study_mins',
       'physical_activity_days', 'sleep_hours', 'sleep_quality',
       'anxiety_score', 'depression_score', 'stress_level', 'happiness_score',
       'focus_score', 'high_risk_flag', 'device_type', 'productivity_score',
       'digital_dependence_score'],
      dtype='str')

Index:
RangeIndex(start=0, stop=3500, step=1)

Statistical summary (Numerical):


,id,age,device_hours_per_day,phone_unlocks,notifications_per_day,social_media_mins,study_mins,physical_activity_days,sleep_hours,sleep_quality,anxiety_score,depression_score,stress_level,happiness_score,focus_score,high_risk_flag,productivity_score,digital_dependence_score
count,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000,3500.000000
mean,1750.500000,28.078571,7.317677,147.107429,335.117143,159.362857,108.116571,3.350000,7.254481,2.708848,7.183858,8.830000,5.076560,6.418568,41.598600,0.201429,65.299301,36.684152
std,1010.507298,9.352385,3.239838,67.770656,239.678612,132.748048,79.796273,1.877976,1.291879,1.101105,5.043220,5.738959,3.403503,3.019828,23.764574,0.401125,9.664716,14.115504
min,1.000000,13.000000,0.280000,9.000000,22.000000,0.000000,0.000000,0.000000,3.000000,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,33.000000,5.600000
25%,875.750000,21.000000,4.867500,96.000000,170.000000,68.000000,42.000000,2.000000,6.410545,1.919438,4.000000,4.000000,1.341488,4.000000,24.750000,0.000000,59.000000,26.200000
50%,1750.500000,27.000000,6.760000,136.000000,268.000000,119.000000,102.000000,3.000000,7.291580,2.907819,6.000000,8.000000,4.809726,7.500000,44.000000,0.000000,65.299301,35.300000
75%,2625.250000,34.000000,9.152500,187.000000,434.000000,204.000000,163.000000,5.000000,8.162298,3.276968,8.845008,13.000000,8.794302,8.900000,59.000000,0.000000,71.000000,45.100000
max,3500.000000,50.000000,17.160000,374.000000,1211.000000,631.000000,418.000000,7.000000,11.004566,5.000000,27.154217,27.000000,10.000000,10.000000,100.000000,1.000000,95.000000,89.200000



Dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        3500 non-null   int64  
 1   age                       3500 non-null   int64  
 2   gender                    3500 non-null   str    
 3   region                    3500 non-null   str    
 4   income_level              3500 non-null   str    
 5   education_level           3500 non-null   str    
 6   daily_role                3500 non-null   str    
 7   device_hours_per_day      3500 non-null   float64
 8   phone_unlocks             3500 non-null   int64  
 9   notifications_per_day     3500 non-null   int64  
 10  social_media_mins         3500 non-null   int64  
 11  study_mins                3500 non-null   int64  
 12  physical_activity_days    3500 non-null   float64
 13  sleep_hours               3500 non-null   float64
 1

id                          0
age                         0
gender                      0
region                      0
income_level                0
education_level             0
daily_role                  0
device_hours_per_day        0
phone_unlocks               0
notifications_per_day       0
social_media_mins           0
study_mins                  0
physical_activity_days      0
sleep_hours                 0
sleep_quality               0
anxiety_score               0
depression_score            0
stress_level                0
happiness_score             0
focus_score                 0
high_risk_flag              0
device_type                 0
productivity_score          0
digital_dependence_score    0
dtype: int64


Duplicate rows:
0


### Summary of the Dataset

From the initial exploration, the dataset contains **3500 rows** and **24 columns**, meaning there are 3500 observations and 24 features describing digital lifestyle, personal characteristics, and mental well-being.


### Dataset Structure

| Category              | Details |
|----------------------|--------|
| Number of Rows       | 3500 |
| Number of Columns    | 24 |
| Numerical Features   | 18 |
| Categorical Features | 6 |
| Missing Values       | None |
| Duplicate Rows       | 0 |

The dataset includes both **numerical** and **categorical** variables. Numerical features represent measurable values such as screen time, sleep hours, and scores, while categorical features represent grouped information such as gender, region, and device type.


### Feature Types

- **Numerical Features**:
  - age
  - device_hours_per_day
  - phone_unlocks
  - notifications_per_day
  - social_media_mins
  - study_mins
  - sleep_hours
  - stress_level
  - productivity_score
  - and other score-based variables

- **Categorical Features**:
  - gender
  - region
  - income_level
  - education_level
  - daily_role
  - device_type


### Key Observations

- The dataset is **complete**, with no missing values in any column.
- There are **no duplicate rows**, meaning each record is unique.
- The data types are correctly separated into:
  - numerical (`int64`, `float64`)
  - categorical (`str`)
- The dataset covers a wide range of features related to:
  - digital usage behavior
  - mental health indicators
  - productivity and lifestyle factors



### Statistical Insights

| Feature                    | Mean | Min | Max |
|---------------------------|------|-----|-----|
| Age                       | 28.08 | 13 | 50 |
| Device Hours per Day      | 7.32 | 0.28 | 17.16 |
| Phone Unlocks             | 147 | 9 | 374 |
| Notifications per Day     | 335 | 22 | 1211 |
| Sleep Hours               | 7.25 | 3 | 11 |
| Stress Level              | 5.08 | 1 | 10 |
| Productivity Score        | 65.30 | 33 | 95 |
| Digital Dependence Score  | 36.68 | 5.6 | 89.2 |

These statistics indicate:
- Users spend an average of **7.3 hours on devices daily**
- Sleep averages around **7.25 hours**
- Stress levels range from **low to high (1–10 scale)**
- Productivity scores are moderately distributed



## 4. Data Cleaning

### 4.1 Handling Missing Values

All columns in the dataset contain **0 missing values**, which means the dataset is complete.

No imputation or removal is required for missing data.

In [5]:
# Check missing values in each column
print("Missing values:")
display(df.isnull().sum())

Missing values:


id                          0
age                         0
gender                      0
region                      0
income_level                0
education_level             0
daily_role                  0
device_hours_per_day        0
phone_unlocks               0
notifications_per_day       0
social_media_mins           0
study_mins                  0
physical_activity_days      0
sleep_hours                 0
sleep_quality               0
anxiety_score               0
depression_score            0
stress_level                0
happiness_score             0
focus_score                 0
high_risk_flag              0
device_type                 0
productivity_score          0
digital_dependence_score    0
dtype: int64

### 4.2 Handling Duplicate Records

There are **no duplicate rows** in the dataset.

This ensures that each record is unique and no bias is introduced from repeated data.

In [6]:
# Check duplicate rows
print("Duplicate rows:")
print(df.duplicated().sum())

Duplicate rows:
0


### 4.3 Data Type Verification

The dataset contains:
- numerical data (int64, float64)
- categorical data (string)

The data types appear correct and suitable for analysis. No conversion is required at this stage.

In [7]:
# Check data types
print("Data types:")
df.dtypes

Data types:


id                            int64
age                           int64
gender                          str
region                          str
income_level                    str
education_level                 str
daily_role                      str
device_hours_per_day        float64
phone_unlocks                 int64
notifications_per_day         int64
social_media_mins             int64
study_mins                    int64
physical_activity_days      float64
sleep_hours                 float64
sleep_quality               float64
anxiety_score               float64
depression_score            float64
stress_level                float64
happiness_score             float64
focus_score                 float64
high_risk_flag                int64
device_type                     str
productivity_score          float64
digital_dependence_score    float64
dtype: object

### 4.4 Check Logical Consistency

we divide it into two
- Categorical feature(Text, String)
- Numerical feature 

In [8]:
# categorical feature

# check unique value 
for col in ['gender','region','income_level','education_level','daily_role','device_type']:
    print(col, df[col].unique())

# check inconsistent format 
for col in ['gender','region','income_level','education_level','daily_role','device_type']:
    df[col] = df[col].str.strip().str.lower()

gender <StringArray>
['Female', 'Male']
Length: 2, dtype: str
region <StringArray>
['Asia', 'Africa', 'North America', 'Middle East', 'Europe', 'South America']
Length: 6, dtype: str
income_level <StringArray>
['High', 'Lower-Mid', 'Low', 'Upper-Mid']
Length: 4, dtype: str
education_level <StringArray>
['High School', 'Master', 'Bachelor', 'PhD']
Length: 4, dtype: str
daily_role <StringArray>
[   'Part-time/Shift', 'Full-time Employee',     'Caregiver/Home',
 'Unemployed_Looking',            'Student']
Length: 5, dtype: str
device_type <StringArray>
['Android', 'Laptop', 'Tablet', 'iPhone']
Length: 4, dtype: str


In [9]:
# numerical feature

# checking hours: hours cannot exceed 24 hours and could not be negative value 
(df['device_hours_per_day'] < 0).sum()
(df['device_hours_per_day'] > 24).sum()
(df['notifications_per_day'] < 0).sum()
(df['social_media_mins'] < 0).sum()
(df['study_mins'] < 0).sum()
(df['sleep_hours'] < 0).sum()
(df['sleep_hours'] > 24).sum()

# checking day should be in between 0 to 7
(df['physical_activity_days'] < 0).sum()
(df['physical_activity_days'] > 7).sum()

# score should be in range 0 - 10
(df['stress_level'] < 0).sum()
(df['stress_level'] > 10).sum()

# productivity should be in range 0 - 100
(df['productivity_score'] < 0).sum()
(df['productivity_score'] > 100).sum()


np.int64(0)

### 4.5 remove some features that are not important

Several features were removed to improve model performance and avoid data leakage.

- The id column was removed as it does not provide useful information.
- Anxiety, depression, and happiness scores were removed because they are strongly related to stress level.
- Focus score was removed as it directly relates to productivity.
- High risk flag was removed as it may be derived from stress-related features.

The remaining features include relevant behavioral and contextual variables for prediction.

In [10]:
df = df.drop(columns=[
    'id',
    'region',
    'income_level',
    'daily_role',
    'anxiety_score',
    'device_type',
    'depression_score',
    'happiness_score',
    'focus_score',
    'high_risk_flag',
    'digital_dependence_score'
])

### 4.6 Outlier Detection

In this step, we detect outliers in numerical features.

Outliers are extreme values that differ significantly from other observations.

They may represent:
- real but rare behavior
- data entry errors

Outliers are detected using the IQR (Interquartile Range) method.

In [11]:
# Select numerical columns
num_cols = df.select_dtypes(include=['int64', 'float64']).columns

# Function to count outliers
def count_outliers(df, col):
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    return ((df[col] < lower) | (df[col] > upper)).sum()

# Check outliers for each column
for col in num_cols:
    print(f"{col}: {count_outliers(df, col)} outliers")

age: 0 outliers
device_hours_per_day: 32 outliers
phone_unlocks: 16 outliers
notifications_per_day: 189 outliers
social_media_mins: 238 outliers
study_mins: 11 outliers
physical_activity_days: 0 outliers
sleep_hours: 20 outliers
sleep_quality: 0 outliers
stress_level: 0 outliers
productivity_score: 52 outliers


In [12]:
# Remove impossible values only

df = df[df['device_hours_per_day'] <= 24]
df = df[df['sleep_hours'] <= 24]
df = df[df['physical_activity_days'] <= 7]

### Outlier Detection Summary

Outliers were identified using the IQR method across numerical features.

| Feature                   | Outliers |
|---------------------------|----------|
| device_hours_per_day      | 32 |
| phone_unlocks             | 16 |
| notifications_per_day     | 189 |
| social_media_mins         | 238 |
| study_mins                | 11 |
| sleep_hours               | 20 |
| productivity_score        | 52 |
| digital_dependence_score  | 33 |

Some features such as age, physical activity days, sleep quality, and stress level showed no outliers.


### Interpretation & Decision

Outliers are mainly found in behavioral features such as device usage and social media activity.

These values are **not removed** because:
- they represent **realistic human behavior**
- they provide **important signals for the model**
- no invalid or impossible values were detected

Removing them may lead to loss of meaningful information and reduce model performance.


## Copy clean data to csv

In [13]:
# Save cleaned dataset
df.to_csv("../00-dataSource/cleaned_data.csv", index=False)

print("Cleaned dataset saved successfully!")


Cleaned dataset saved successfully!
